lib/spack/llnl/util/tty/pty.py

```diff
@@ -0,0 +1,287 @@
+        return status[1]
+
+    def wait_stopped(self):
+        self.wait(lambda: "T" in self.proc_status())
+
+    def wait_running(self):
+        status = self.proc_status()
+        self.wait(lambda: "T" not in status)
+
+
+def _child_process(
+        tty_name, stdout_fd, stderr_fd, ready, child_function, attrs):
+    """Child process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``child_function``.
+
+    """
+    # new process group, like a command or pipeline launched by a shell
+    os.setpgrp()
+
+    # take controlling terminal and set up pty IO
+    stdin_fd = os.open(tty_name, os.O_RDWR)
+    os.dup2(stdin_fd, sys.stdin.fileno())
+    os.dup2(stdout_fd, sys.stdout.fileno())
+    os.dup2(stderr_fd, sys.stderr.fileno())
+    os.close(stdin_fd)
+
+    if attrs.get("debug"):
+        sys.stderr.write("child: stdin.isatty(): %s\n" % sys.stdin.isatty())
+
+    # tell the parent that we're really running
+    if attrs.get("debug"):
+        sys.stderr.write("child: ready!\n")
+    ready.value = True
+
+    child_function(attrs)
+
+
+def _master_process(master_function, child_function, attrs):
+    """Master process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``master_function``.
+
+    """
+    os.setsid()   # new session; this process is the controller
+
+    master_fd, child_fd = os.openpty()
+    pty_name = os.ttyname(child_fd)
+
+    # take controlling terminal
+    pty_fd = os.open(pty_name, os.O_RDWR)
+    os.close(pty_fd)
+
+    ready = multiprocessing.Value('i', False)
+    proc = multiprocessing.Process(
+        target=_child_process,
+        args=(pty_name, sys.stdout.fileno(), sys.stderr.fileno(),
+              ready, child_function, attrs)
+    )
+    proc.start()
+
+    # wait for subprocess to be running and connected.
+    while not ready.value:
+        time.sleep(1e-5)
+        pass
+
+    if attrs.get("debug"):
+        sys.stderr.write("pid:        %d\n" % os.getpid())
+        sys.stderr.write("pgid:       %d\n" % os.getpgrp())
+        sys.stderr.write("sid:        %d\n" % os.getsid(0))
+        sys.stderr.write("tcgetpgrp:  %d\n" % os.tcgetpgrp(master_fd))
+        sys.stderr.write("\n")
+
+        proc_pgid = os.getpgid(proc.pid)
+        sys.stderr.write("child pid:  %d\n" % proc.pid)
+        sys.stderr.write("child pgid: %d\n" % proc_pgid)
+        sys.stderr.write("child sid:  %d\n" % os.getsid(proc.pid))
+        sys.stderr.write("\n")
+        sys.stderr.flush()
+    # set up master to ignore SIGTSTP, like a shell
+    signal.signal(signal.SIGTSTP, signal.SIG_IGN)
+
+    # call the master function once the child is ready
+    try:
+        controller = ProcessController(
+            proc.pid, master_fd, debug=attrs.get("debug"))
+        error = master_function(proc, controller, attrs)
+    except BaseException:
+        error = 1
+        traceback.print_exc()
+
+    proc.join()
+
+    return error or proc.exitcode
```

Comment:
Can you update the docstring to mention what is being returned? I think the idea is that you are returning whether the parent or the child process failed.

Reply:
done